### PR TITLE
Addon-docs: MDX2 support

### PR DESCRIPTION
**Pull request by shilman**
Issue: #17514 #17137

## What I did

- [x] Extract MDXv1 compiler into its own package & clean up snapshot tests
- [x] Create MDXv2 compiler in its own package
- [x] Add opt-in mechanism
- [x] Add MIGRATION docs

## How to test

- [ ] CI passes
- [x] Battery of snapshot tests

### CODE REVIEW

File: addons/docs/src/frameworks/common/preset.ts [hunk 1]
```diff
@@ -62,9 +63,17 @@ export async function webpack(
   } = options;
 
   const mdxLoaderOptions = {
+    skipCsf: true,
     remarkPlugins: [remarkSlug, remarkExternalLinks],
   };
 
+  const mdxVersion = global.FEATURES?.previewMdx2 ? 'MDX2' : 'MDX1';
+  log(`Addon-docs: using ${mdxVersion}`);
```
Comment (tmeasday) [inline addons/docs/src/frameworks/common/preset.ts:71]: Console log is intentional?

**Review by tmeasday — COMMENTED** [review]
Is there a PR for the two packages we should be looking at?

So this means that all apps will depend on `@storybook/mdx1-csf` and folks that opt in will depend on both? Is that going to be OK from a dependencies point of view?

**Comment by shilman** [comment]
@tmeasday Yes, there are two more PRs here

- https://github.com/storybookjs/mdx1-csf/pull/1
- https://github.com/storybookjs/mdx2-csf/pull/1

The dependency structure is as you described. I don't anticipate any dependency conflicts inside Storybook because each package depends on a specific version and Storybook only depends on the packages. It could, however, cause conflicts with the user's app. This is why we've been prebundling dependencies, and I supppose we could consider prebundling these packages as well. This would require @ndelangen to generalize his prebundling script to be used outside the SB monorepo, and I don't think we need to do it right away.

There is one "hack" which is that SB depends on `@mdx-js/react@1.x`, which coincidentally works correctly with v2. We could take this PR a step further and also move that dependency into the new packages. However, I didn't do that because:
1. It would complicate things
2. I hope this split will be temporary until 7.0, at which point we'll upgrade to MDXv2 completely

**Comment by shilman** [comment]
@tmeasday @yannbf Merging this to keep things moving forward & reduce merge conflicts. I'll update the dependencies once the `mdx-csf` PRs are reviewed/merged.

**Comment by agrohs** [comment]
@shilman - wondering if this is documented somewhere around how to use the experimental config w/ mdx2 in storybook 6.5 by chance??

(I'm in a large mono repo that across the board moved to mdx2 and really cannot yet move to storybook 7, so hopeful to get this to run/kick some tires on it! Much appreciate in advance for any help/guidance on it)

**Pull request merged**